ca/server.go

```diff
@@ -808,3 +826,162 @@ func isFinalState(status api.IssuanceStatus) bool {
 
 	return false
 }
+
+func (s *Server) finishRootRotation(expectedRootCA *api.RootCA) error {
+	// when the nodes are all converged on having TLS certificates issued by the correct entity, we can
+	// try to finish the root rotation
+	clusterID := s.securityConfig.ClientTLSCreds.Organization()
+	return s.store.Update(func(tx store.Tx) error {
+		cluster := store.GetCluster(tx, clusterID)
+		if cluster == nil {
+			return fmt.Errorf("unable to get cluster %s", clusterID)
+		}
+
+		// If the RootCA object has changed (because another root rotation was started or because some other node
+		// had finished the root rotation), we cannot finish the root rotation that we were working on.
+		if !equality.RootCAEqualStable(expectedRootCA, &cluster.RootCA) {
+			return errors.New("target root rotation has changed - unable to complete previous root rotation")
+		}
+
+		var signerCert []byte
+		if len(cluster.RootCA.RootRotation.CAKey) > 0 {
```

Comment:
I get it now. I had missed the implications of the `RootCAEqualStable` check. I think this is fine as-is.